test/generators/gen_link/gen_link.ml

```diff
@@ -0,0 +1,235 @@
+      List [ Atom "deps"; Atom (to_string path) ];
+      List
+        [
+          Atom "action";
+          List
+            [
+              Atom "run";
+              Atom "odoc";
+              Atom "compile";
+              Atom "--pkg";
+              Atom "test";
+              Atom "-o";
+              Atom "%{target}";
+              Atom "%{deps}";
+            ];
+        ];
+    ]
+
+let mld_odocl_target_rule path =
+  let open Fpath in
+  let basename = basename (rem_ext path) in
+  List
+    [
+      Atom "rule";
+      List [ Atom "target"; Atom ("page-" ^ basename ^ ".odocl") ];
+      List [ Atom "deps"; Atom ("page-" ^ basename ^ ".odoc") ];
+      List
+        [
+          Atom "action";
+          List
+            [
+              Atom "run";
+              Atom "odoc";
+              Atom "link";
+              Atom "-o";
+              Atom "%{target}";
+              Atom "%{deps}";
+            ];
+        ];
+    ]
+
+let mld_file_rule path =
+  let open Printf in
+  printf "%s\n\n" (to_string_hum (mld_odoc_target_rule path));
+  printf "%s\n\n" (to_string_hum (mld_odocl_target_rule path))
+
+let die s =
+  prerr_endline s;
+  exit 1
+
+let path' () f = Filename.quote (Fpath.to_string f)
+
+let ext' () f = Filename.quote (Fpath.get_ext f)
+
+let gen_rule_for_source_file path =
+  let ext = Fpath.get_ext path in
+  match ext with
+  | ".ml" -> file_rule path ".cmt"
+  | ".mli" -> file_rule path ".cmti"
+  | ".mld" -> mld_file_rule path
+  | _ ->
+      die
+        (sprintf
+           "Don't know what to do with %a because of unrecognized %a extension."
+           path' path ext' path)
+
+let odocls paths =
+  paths
+  |> List.map (fun p ->
+         let open Fpath in
+         let path = relativize ~root:(v "html") p in
+         match path with
+         | Some p -> sprintf "%%{dep:%s}" (to_string p)
+         | None -> assert false)
+  |> List.map (fun p -> Atom p)
+
+let gen_rule_html paths =
+  List
+    [
+      Atom "subdir";
+      Atom "html";
+      List
+        [
+          Atom "rule";
+          List
+            [
+              Atom "with-stdout-to";
+              Atom "html.dune.inc.gen";
+              List
+                [
+                  Atom "pipe-stdout";
+                  List
+                    (Atom "run" :: Atom "gen_html/gen_html.exe" :: odocls paths);
+                  List [ Atom "run"; Atom "dune"; Atom "format-dune-file" ];
+                ];
+            ];
+        ];
+    ]
+
+let diff_rule_html =
+  List
+    [
+      Atom "rule";
+      List [ Atom "alias"; Atom "runtest" ];
+      List
+        [
+          Atom "action";
+          List
+            [
+              Atom "diff";
+              Atom "html/html.dune.inc";
+              Atom "html/html.dune.inc.gen";
+            ];
+        ];
+    ]
+
+let gen_html_rule paths =
+  printf "%s\n\n" (to_string_hum (gen_rule_html paths));
+  printf "%s\n\n" (to_string_hum diff_rule_html)
+
+let read_file_from_dir dir =
```

Comment:
This function should be splitted into "read_file_from_dir" (returns the `paths` list) and "gen_rules" (taking the `paths` list).

Reply:
done